HealthcareSystem/IntegrationAdapters/Controllers/SftpController.cs

```diff
@@ -0,0 +1,49 @@
+﻿using Backend.Service.DrugConsumptionService;
+using Backend.Service.SftpService;
+using Microsoft.AspNetCore.Mvc;
+using Newtonsoft.Json;
+using System;
+using Backend.Model.Pharmacies;
+
+namespace IntegrationAdapters.Controllers
+{
+    public class SftpController : Microsoft.AspNetCore.Mvc.Controller
```

Comment:
So the use case here is just sending a consumption report to a pharmacy, right? I feel like there should be a bit more abstraction happening here, I don't like the fact that we're doing something via SFTP reaching all the way to the controller since I assume it's just an implementation detail. Think something like, IDrugConsumptionReportService (bad name probably, but again, I don't know the exact use case). And then a specific implementation of that Service that happens to use the SftpService (really, SftpCommunicator) to send out the reports.

Reply:
Report sending was just the first use of SFTP that was implemented. As far as I know, this sprint involves implementations of other use cases requiring SFTP, so I think the controller is fine as it is right now, especially considering we are still not sure how the frontend will look in the end.

Reply:
Okay, I'm alright with leaving it as is for now then, although the fact that SFTP will have other uses here actually reinforces the point that it shouldn't be used by the controllers directly. So, as I understand, the exact use case here is approximately this: user chooses the dates and hits the send report button, the report is generated, the generated report is then sent to the pharmacy (or all pharmacies?). 
So here is a really long and rambling discussion of the layer structure for this:
- The Controller has a method that takes the dates and maybe like the pharmacy id if that's necessary. It passes all of that to the corresponding Service method, and just deals with things like catching exceptions, converting error messages, and such (although I'm not sure how any of that goes with MVC and how necessary it is, I'm much better acquainted with REST Controllers). In REST such a Controller would also be easily testable because you could just make a bunch of stubs of the Service that generate appropriate exceptions, and then check if the return messages you're getting are correct. This is probably not the case with MVC though, since there's redirection to the View happening and such?
- The Service also has a method that directly corresponds to the use case, so it also takes the dates and possibly pharmacy id. The Service then orchestrates between the other parts of the system, so here it would do something like call the Repository to get the data for the report based on the dates, possibly also call another Repository to get data needed for communication with the specific pharmacy by the id (or maybe communication data for all pharmacies if that's what we're doing instead), and then call the Communication layer to actually send the message.
- So, the Communication layer, or whatever it ends up being called. I would actually make that have a domain level abstraction at the very top as well, not just the generic classes for specific protocols. The reason for that is really just to prevent changes to the Service layers when the protocol changes. Actually, here's a thing I just realized, isn't this particular domain object or set of domain objects exactly what the goal of the project here is though? I feel like what I'm talking about here would actually just be the famous IPharmacyAdapter interface or something? Anyway, the point is that for this use case this particular class would have a method that's really just something like SendReports which takes a list of the DrugConsuptionReport objects which the Service fetched, and also gets the data needed for communication somehow (either the Service also fetched that and injects it here, or it was already in this class somehow, I really haven't worked what would be good for that and what wouldn't but I think it'll get clearer after we learn DDD). This way the Service isn't dealing with forming a specific message from the report list or anything, and it's nicely abstracted away from any communication protocols. As a consequence, the Service or Service and Repository together or whatever are also super easy to test, cause you can just mock the IPharmacyAdapter interface and check if the correct method gets called with the correct parameters.  And the method in a particular implementation of IPharmacyAdapter or whatever it's called would then form a protocol appropriate message and delegate sending it to the generic ISftpCommunicator or whatever. Now that I think about it, this is actually kind of similar to the discussion about loggers and domain loggers in particular what we needed to read for last weeks questionnaire? Any kind of message sending ends up being similar in design I suppose.

Anyway, obviously many things still need to be worked out better here, but I hope it gives you some useful ideas.

Reply:
I appreciate the input and ideas. I think this will require a lot of designing as the whole team, but for now I would leave it like this so everyone can start working on their features and later on when we properly design everything I will just refactor it.

Reply:
Yeah, I agree, I would also leave it as is for now, I just thought it was a good starter for a discussion.